app/services/region_report_service.rb

```diff
@@ -23,20 +27,71 @@ def initialize(region:, period:, current_user:)
   attr_reader :facilities
   attr_reader :organizations
   attr_reader :period
+  attr_reader :range
   attr_reader :region
 
   def call
     compile_control_and_registration_data
     data.merge! compile_cohort_trend_data
+    data[:visited_without_bp_taken] = count_visited_without_bp_taken
+    data[:visited_without_bp_taken_rate] = percentage_visited_without_bp_taken
+    data[:missed_visits] = calculate_missed_visits
+    data[:missed_visits_rate] = calculate_missed_visits_rate
     data[:top_region_benchmarks].merge!(top_region_benchmarks)
 
     data
   end
 
+  def calculate_missed_visits
+    data[:cumulative_registrations].each_with_object({}) do |(period, count), hsh|
+      hsh[period] = count - data[:controlled_patients][period] - data[:uncontrolled_patients][period]
+    end
+  end
+
+  def calculate_missed_visits_rate
+    data[:missed_visits].each_with_object({}) do |(period, count), hsh|
+      hsh[period] = percentage(count, data[:cumulative_registrations][period].to_f)
+    end
+  end
+
+  # visited in last 3 months but had no BP taken
+  def count_visited_without_bp_taken
+    periods = data[:registrations].keys
+    periods.each_with_object({}) do |period, hsh|
+      visits_without_bp = Appointment.from(patients_visited_without_bp_taken(period), "appointments").count("appointments.patient_id")
+      hsh[period] = visits_without_bp
+    end
+  end
+
+  def percentage_visited_without_bp_taken
+    data[:visited_without_bp_taken].each_with_object({}) do |(period, count), hsh|
+      hsh[period] = if count&.zero?
+        0
+      else
+        percentage(count, data[:cumulative_registrations].fetch(period))
+      end
+    end
+  end
+
+  def percentage(numerator, denominator)
+    return 0 if denominator == 0
+    ((numerator.to_f / denominator) * 100).round(0)
+  end
+
+  def patients_visited_without_bp_taken(period)
+    end_date = period.to_date
+    begin_date = period.advance(months: -3).to_date
+    Appointment.select("DISTINCT ON (appointments.patient_id) appointments.*")
+      .joins(:patient).merge(Patient.with_hypertension)
+      .joins("LEFT OUTER JOIN blood_pressures ON blood_pressures.patient_id = appointments.patient_id")
+      .order("appointments.patient_id")
+      .where("blood_pressures.id IS NULL")
+      .where("appointments.device_created_at :: DATE BETWEEN :begin AND :end", begin: begin_date, end: end_date)
+      .where("patients.registration_facility_id in (?)", facilities)
```

Comment:
@simpledotorg/server-developers how does this query look for patients who have a visit but no BP taken?

Reply:
This looks correct mostly, but I feel like this:

```
where("patients.registration_facility_id in (?)", facilities)
```

Should instead be on the **creation** facility of the appointment instead?

It'd make sense to say that a patient is counted as "visited" if they have an appointment created at any of the facilities in the Region.

wdyt?

Also, we're not explicitly checking for prescription drugs, are we using appointments as proxy for meds updated? If so, I think that seems sensible.

Reply:
My interpretation was that the facility scoping here is the "assigned facility". I don't care where you scheduled your appointment, I care that the patient I was responsible for had an appointment scheduled, aka, they visited any clinic. I suspect this will become `patients.assigned_facility_id in (?)` soon.

Reply:
👍